lua/starfall/libs_cl/bass.lua

```diff
@@ -1,20 +1,107 @@
 
 local plyCount = SF.LimitObject("bass", "bass sounds", 20, "The number of sounds allowed to be playing via Starfall client at once")
+local soundDatas = {} -- { [IGModAudioChannel] = { ... }, ... } -- Contains extra data for each starfall sound.
+local simpleFadeSounds = {} -- { [1] = IGModAudioChannel, ... } -- A list of sounds that need to be manually controlled through 'simple fading'.
 
 SF.ResourceCounters.Bass = {icon = "icon16/sound_add.png", count = function(ply) return plyCount:get(ply).val end}
 
-local function deleteSound(ply, sound)
-	if sound:IsValid() then sound:Stop() end
+
+-- Calculates whether or not a 3D sound is out of earshot, and its fade multiplier.
+local function getSimpleFading(snd)
+	local pos = snd:GetPos()
+	local earPos = EyePos()
+	local fadeMin = snd:Get3DFadeDistance()
+	local fadeMax = soundDatas[snd].fadeMax
+	local distSqr = pos:DistToSqr(earPos)
+
+	if distSqr <= fadeMin * fadeMin then return false, 1 end
+	if distSqr >= fadeMax * fadeMax then return true, 0 end
+
+	local dist = math_sqrt(distSqr)
+	local farnessFrac = (dist - fadeMin) / (fadeMax - fadeMin)
+
+	return false, (1 - farnessFrac) ^ 2 -- Sounds falls off with dist^2. Unfortunately, we still have to do the square root inbetween.
+end
+
+local function applySimpleFading(snd)
+	local sndData = soundDatas[snd]
+	local outOfEarshot, fadeMult = getSimpleFading(snd)
+
+	if sndData.fadeMult ~= fadeMult then
+		sndData.fadeMult = fadeMult
+		sndData.outOfEarshot = outOfEarshot
+
+		snd:SetVolume(sndData.targetVolume * fadeMult)
+	end
+end
+
+local function addSoundToSimpleFade(snd)
+	local sndData = soundDatas[snd]
+	if sndData.simpleFadeEnabled then return end
+
+	sndData.simpleFadeEnabled = true
+	table.insert(simpleFadeSounds, snd)
+	applySimpleFading(snd)
+
+	if #simpleFadeSounds ~= 1 then return end -- Hook is already running.
+
+	hook.Add("Think", "SF_Bass_SimpleFade", function()
+		for _, s in ipairs(simpleFadeSounds) do
+			if s:IsValid() and s:GetState() == GMOD_CHANNEL_PLAYING then
+				applySimpleFading(s)
+			end
+		end
+	end)
+end
+
+local function removeSoundFromSimpleFade(snd)
+	local sndData = soundDatas[snd]
+	if not sndData.simpleFadeEnabled then return end
+
+	sndData.simpleFadeEnabled = false
+	sndData.fadeMult = 1
+	sndData.outOfEarshot = false
+	table.RemoveByValue(simpleFadeSounds, snd)
+
+	if snd:IsValid() then -- Could be in the process of being deleted.
+		snd:SetVolume(sndData.targetVolume) -- Remove manual fading, reset to target volume.
+	end
+
+	if #simpleFadeSounds ~= 0 then return end -- Hook is still needed
+
+	hook.Remove("Think", "SF_Bass_SimpleFade")
+end
+
+local function deleteSound(ply, snd)
+	if snd:IsValid() then snd:Stop() end
+	if not soundDatas[snd] then return end
+
+	removeSoundFromSimpleFade(snd)
+	soundDatas[snd] = nil
 	plyCount:free(ply, 1)
 end
 
+-- Sets the min/max fade distance of a sound, and whether or not to use simple fading.
+local function setSoundFade(snd, min, max, useSimpleFading)
```

Comment:
This function is pointless. You could easily move the Set3DFadeDistance to addSoundToSimpleFade/removeSoundFromSimpleFade using soundData.fadeMin and soundData.fadeMax for removeSoundFromSimpleFade.

The remaining `soundData.fadeMax = max, soundData.fadeMin = min` can go in bass_methods:setFade

Reply:
Ultimately, the only way to make add/remove SimpleFade set the internal fade distance would require either turning `:setFade()` into an ugly mess (see below) or passing `min, max` into their arguments which is unclean, especially since `removeSoundFromSimpleFade()` is also called by `destroySound()`

![image](https://github.com/thegrb93/StarfallEx/assets/26103433/977a5621-5fd5-4a83-ae7a-2c46c28f5e71)

This could be made less ugly by making each block do `min = ...`, `max = ...` and then apply to `sndData`, but the extra lines would lead to the full change being a net nothing in terms of line count.

Removing `setSoundFade()` and splitting it apart between `:setFade()` and the add/remove functions just complicates the code further. If you tuck it all into `:setFade()` instead, then we're back to sound init needing to either forcibly call `:setFade()` on the instance, or set the fade distance manually, neither of which are clean.

Reply:
What we have now seems clean enough now. Thanks. I might do another PR simplify it a bit